Add CI cron jobs

TerrenceMcGuinness-NOAA opened the pull request:
**Description**

As a maintainer of the CI framework, I need a set of cron jobs that will fully automate the CI pipeline so that whenever the appropriate label on GitHub is created the PR gets cloned and built followed by a set of functional experiments that are executed and reported on.

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

**How Has This Been Tested?**

- [x] Clone and Build tests on Orion and Hera
- [x] Cycled test on Orion
- [x] Cycled test on Hera
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

TerrenceMcGuinness-NOAA committed:
disambiguated env vars for HOMEgfs and GFS_CI_ROOT and removed re-assigned and redundent variables

TerrenceMcGuinness-NOAA committed:
debugged updates to scripts (typos ect)

TerrenceMcGuinness-NOAA committed:
Update create_experiment.py

TerrenceMcGuinness-NOAA committed:
Update driver.sh

TerrenceMcGuinness-NOAA committed:
Update clone-build_ci.sh

TerrenceMcGuinness-NOAA committed:
added another layer in file structure for experments where each experment has a dir that contains COMROT and config

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
fixed bugs with tree restructure under ../PR/{PR#}/RUNTESTS

TerrenceMcGuinness-NOAA committed:
spelling errors in create experiment

TerrenceMcGuinness-NOAA committed:
added run_ci.sh that is used in cron to run expirments

TerrenceMcGuinness-NOAA committed:
removed --login from shabang in bash scripts

Commit:
added environment for orion

Commit:
fixed bug with needing updated git from module load on orion

TerrenceMcGuinness-NOAA committed:
added correct ICDIR to orion and module load module_ci for uniformally getting rocoto for run_ci.sh

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
added old python compair GFS folders for development

TerrenceMcGuinness-NOAA committed:
added Execute class for nccnp to comparitor

TerrenceMcGuinness-NOAA committed:
udpate logging level and fixed spelling errors

TerrenceMcGuinness-NOAA committed:
added not none over len 0 for support Null return on Execute on nccmp

TerrenceMcGuinness-NOAA committed:
cleaned up and and made outout human readable for the comaritor code

TerrenceMcGuinness-NOAA committed:
update GitHub labels to all start with CI

TerrenceMcGuinness-NOAA committed:
format output dropped f and added nccmp to orion env modufile

TerrenceMcGuinness-NOAA committed:
added grib file checker with corrlation checks in compare folders script

TerrenceMcGuinness-NOAA committed:
got wgrib2 checker to work

TerrenceMcGuinness-NOAA committed:
added wgrib2 to ci module file for Hera

TerrenceMcGuinness-NOAA committed:
had to add hpc 1.1.0 to module file to get wgrib2

TerrenceMcGuinness-NOAA committed:
fixed module load order for virtutal python environment

TerrenceMcGuinness-NOAA committed:
started code cleanup, removed job switches for storing a yaml file of all files

TerrenceMcGuinness-NOAA committed:
removed import collections

TerrenceMcGuinness-NOAA committed:
test git edit

TerrenceMcGuinness-NOAA committed:
pcode edits

TerrenceMcGuinness-NOAA committed:
more pycode edits

TerrenceMcGuinness-NOAA committed:
more pycode edits

TerrenceMcGuinness-NOAA committed:
more pycode edits

TerrenceMcGuinness-NOAA committed:
more pycode edits

TerrenceMcGuinness-NOAA committed:
more pycode edits

TerrenceMcGuinness-NOAA committed:
few more pycode edits

TerrenceMcGuinness-NOAA committed:
pycode edits

TerrenceMcGuinness-NOAA committed:
two more pycode edits

TerrenceMcGuinness-NOAA committed:
bash lint edit

TerrenceMcGuinness-NOAA committed:
fixed some formatting issues from pycode updates

TerrenceMcGuinness-NOAA committed:
some more cleaning up format

TerrenceMcGuinness-NOAA committed:
moved module load in driver

TerrenceMcGuinness-NOAA committed:
added ignore list for wild cards

TerrenceMcGuinness-NOAA committed:
removed module load in driver and added rm pr_list at the end if used in cron

TerrenceMcGuinness-NOAA committed:
fix syntax error with echo at end of clone-build

TerrenceMcGuinness-NOAA committed:
need to remove set -e from driver and build scripts to that CI system can report back to GitHub the fails and continue their work on other builds

TerrenceMcGuinness-NOAA committed:
added checks and output results for checkout in clone-build script

TerrenceMcGuinness-NOAA committed:
moved message sent to GitHub on the status of each PR on pass or fail to end of loop to pick up all cases

TerrenceMcGuinness-NOAA committed:
added two seperate module loads for workflow and CI, fixed some messaging

TerrenceMcGuinness-NOAA committed:
added gwci and gwsetup module files for orion

TerrenceMcGuinness-NOAA committed:
update machine gwsetup module load in run_ci script

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
udpated gwci on orion for threaded version of nccmp, updated reporting on grib files in compare, small bug in test/diff_grib with arg array

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature_1428_CI-cronjobs' of github.com:TerrenceMcGuinness-NOAA/global-workflow into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
cleaned up path reporting in compare folders and fixed dependacies in gwci module file for Hera

TerrenceMcGuinness-NOAA committed:
updated verbose output and added ignore list in file count conformity (had extra logs from multiple runs)

TerrenceMcGuinness-NOAA committed:
few pythoncode lint updates

TerrenceMcGuinness-NOAA committed:
more pycode linting (spaces and commas)

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
added || exit on cd (change directory) in clone-build BASH script

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature_1428_CI-cronjobs' of github.com:TerrenceMcGuinness-NOAA/global-workflow into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
indentation for visual effect PIP8

TerrenceMcGuinness-NOAA committed:
indentation for visual effect PIP8 E127, again

TerrenceMcGuinness-NOAA committed:
PIP8 E127 wanted alignment after continuation line to be a regular indent

TerrenceMcGuinness-NOAA committed:
PEP8 black line had spaces

TerrenceMcGuinness-NOAA committed:
type spelling error of directories

WalterKolczynski-NOAA commented on ci/environments/orion.sh:
For a future (very small) PR, can you reorganize one of these so the variables are in the same order as the other?

TerrenceMcGuinness-NOAA replied:
As the other **${MACHINE}.sh** scripts?

aerorahul commented on ci/experiments/C96_atm3DVar_also.yaml:
```diff
@@ -0,0 +1,14 @@
+experiment:
```
what is the difference between this case and `C96_atm3DVar.yaml`?

TerrenceMcGuinness-NOAA replied:
Nothing, it is how I tested multiple experiments with only having one and I use it for the back-to-back validation tests for baselines.

aerorahul replied:
understood.

aerorahul commented on ci/experiments/C96_atm3DVar.yaml:
```diff
@@ -0,0 +1,14 @@
+  app: ATM
+  resdet: 96
+  comrot: ${RUNTESTS}/${pslot}/COMROT
+  expdir: ${RUNTESTS}/${pslot}/EXPERMENT
```
```suggestion
  expdir: ${RUNTESTS}/${pslot}/EXPERIMENT
```

TerrenceMcGuinness-NOAA replied:
F&@$, I keep doing that!

aerorahul commented on ci/scripts/clone-build_ci.sh:
```diff
@@ -1,11 +1,6 @@
 #!/bin/bash
-set -eux
+set -ux
```
I think we want all errors to be trapped. No?
```suggestion
set -eux
```

WalterKolczynski-NOAA replied:
If you fail immediately on error, you can't check for the error and report it to the log (without adding a real bash trap or a script above this one). I think turning off `-e` is the correct move here, just need to be extra careful about checking and handling errors.

aerorahul replied:
good point.  I would think we would disable that (set +e) where we expect to fail.

TerrenceMcGuinness-NOAA replied:
@WalterKolczynski-NOAA Correct!

WalterKolczynski-NOAA replied:
I think Rahul is right, let's just turn off `-e` where we will encounter errors with the PR, like checkout and build (basically where there are already status checks).

TerrenceMcGuinness-NOAA replied:
OK, let me look at that and fine tune it.  I just caught Rahul's `set +e `suggesting, sounds like a plan.

aerorahul commented on ci/scripts/clone-build_ci.sh:
```diff
@@ -73,37 +68,54 @@ if [[ "${pr_state}" != "OPEN" ]]; then
 
 # run build script
-cd sorc
+cd sorc || exit
```
The `exit` should be `non-zero` in case of a failure.
```suggestion
cd sorc || exit 1
```

TerrenceMcGuinness-NOAA replied:
Good one! Pun not intended. ;)

aerorahul commented on ci/scripts/clone-build_ci.sh:
```diff
@@ -49,21 +44,21 @@ done
-git clone "${repo_url}"
-cd global-workflow
+git clone "${REPO_URL}"
+cd global-workflow || exit
```
`exit` should be `non-zero` in case of a failure.
```suggestion
cd global-workflow || exit 1
```

TerrenceMcGuinness-NOAA replied:
Yes!

aerorahul commented on ci/scripts/clone-build_ci.sh:
```diff
@@ -73,37 +68,54 @@ if [[ "${pr_state}" != "OPEN" ]]; then
 rm -rf log.build
-./checkout.sh -g -c
+./checkout.sh -c -g -u &>> log.checkout
+
```
```suggestion
```

aerorahul commented on ci/scripts/clone-build_ci.sh:
```diff
@@ -73,37 +68,54 @@ if [[ "${pr_state}" != "OPEN" ]]; then
+
 # build full cycle
-./build_all.sh -g &>> log.build
+module purge
```
Please use `module-setup.sh` here.  `module purge` might work here, but is not the general solution for all machines.

aerorahul commented on ci/scripts/clone-build_ci.sh:
```diff
@@ -73,37 +68,54 @@ if [[ "${pr_state}" != "OPEN" ]]; then
 fi
 
 ./link_workflow.sh
+build_status=$?
```
This is the return status of the `link_workflow.sh` and not that of the build.
If the build failed from line 102, but the link succeeded, the actual status to be reported is that of the failure.

aerorahul commented on ci/scripts/compare_folders.py:
```diff
@@ -0,0 +1,345 @@
+#!/usr/bin/env python3
```
I am going to ignore reviewing this for now as we are not quite ready to do this.

TerrenceMcGuinness-NOAA replied:
agreed

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -17,65 +18,46 @@
 # Setup the reletive paths to scripts and PS4 for better logging 
 ################################################################
-WF_ROOT_DIR="$(cd "$(dirname  "${BASH_SOURCE[0]}")/../.." >/dev/null 2>&1 && pwd )"
+HOMEGFS_DIR="$(cd "$(dirname  "${BASH_SOURCE[0]}")/../.." >/dev/null 2>&1 && pwd )"
```
```suggestion
HOMEgfs="$(cd "$(dirname  "${BASH_SOURCE[0]}")/../.." >/dev/null 2>&1 && pwd )"
```
There is no need to call the same thing with different names.

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -17,65 +18,46 @@
-fi
-
-case ${TARGET} in
+source "${HOMEGFS_DIR}/ush/detect_machine.sh"
```
```suggestion
source "${HOMEgfs}/ush/detect_machine.sh"
```
Might as well call `${HOMEgfs}/ush/module-setup.sh` here.

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -17,65 +18,46 @@
     exit 1
     ;;
 esac
+export MACHINE_ID
```
```suggestion
```
Isn't this already set?

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -17,65 +18,46 @@
+######################################################
+# setup runtime env for correct python install and git
+######################################################
+module use "${HOMEGFS_DIR}/modulefiles"
```
```suggestion
module use "${HOMEgfs}/modulefiles"
```

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -17,65 +18,46 @@
     ;;
 esac
+export MACHINE_ID
+export REPO_URL=${REPO_URL:-"https://github.com/NOAA-EMC/global-workflow.git"}
```
How about move this to the top after line 21 where `GH` is exported.

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -84,71 +66,57 @@ else
-  id=$("${GH}" pr view "${pr}" --repo "${repo_url}" --json id --jq '.id')
-  "${WF_ROOT_DIR}/ci/scripts/clone-build_ci.sh" -p "${pr}" -d "${pr_dir}" -o "${pr_dir}/output_${id}"
+  id=$("${GH}" pr view "${pr}" --repo "${REPO_URL}" --json id --jq '.id')
+  "${HOMEGFS_DIR}/ci/scripts/clone-build_ci.sh" -p "${pr}" -d "${pr_dir}" -o "${pr_dir}/output_${id}"
```
```suggestion
  "${HOMEgfs}/ci/scripts/clone-build_ci.sh" -p "${pr}" -d "${pr_dir}" -o "${pr_dir}/output_${id}"
```

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -17,65 +18,46 @@
-    echo "Running Automated Testing on ${TARGET}"
-    source "${WF_ROOT_DIR}/ci/environments/${TARGET}.sh"
+    echo "Running Automated Testing on ${MACHINE_ID}"
+    source "${HOMEGFS_DIR}/ci/environments/${MACHINE_ID}.sh"
```
```suggestion
    source "${HOMEgfs}/ci/environments/${MACHINE_ID}.sh"
```

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -84,71 +66,57 @@ else
+    rm -Rf "${RUNTESTS:?}"/*
     #############################################################
-    # loop over every yaml file in ${WF_ROOT_DIR}/ci/experiments
+    # loop over every yaml file in ${HOMEGFS_DIR}/ci/experiments
```
```suggestion
    # loop over every yaml file in ${HOMEgfs}/ci/experiments
```

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -84,71 +66,57 @@ else
     #############################################################
-    for yaml_config in "${WF_ROOT_DIR}/ci/experiments/"*.yaml; do
+    module load "module_gwsetup.${MACHINE_ID}"
+    for yaml_config in "${HOMEGFS_DIR}/ci/experiments/"*.yaml; do
```
```suggestion
    for yaml_config in "${HOMEgfs}/ci/experiments/"*.yaml; do
```

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -84,71 +66,57 @@ else
       pslot=$(basename "${yaml_config}" .yaml) || true
       export pslot
-      "${WF_ROOT_DIR}/ci/scripts/create_experiment.py" --yaml "${WF_ROOT_DIR}/ci/experiments/${pslot}.yaml"
+      "${HOMEGFS_DIR}/ci/scripts/create_experiment.py" --yaml "${HOMEGFS_DIR}/ci/experiments/${pslot}.yaml" --dir "${pr_dir}/global-workflow"
```
```suggestion
      "${HOMEgfs}/ci/scripts/create_experiment.py" --yaml "${HOMEgfs}/ci/experiments/${pslot}.yaml" --dir "${pr_dir}/global-workflow"
```

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -0,0 +1,70 @@
+#
+# Script description: BASH script for checking for experiments in a given PR and 
+#                     simply running rocotorun on each.  This script is intended
+#                     to run from within a cron job in the CI Mangers account
```
```suggestion
#                     to run from within a cron job in the CI Managers account
```

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -0,0 +1,70 @@
+# Abstract TODO
+#####################################################################################
+
+HOMEGFS_DIR="$(cd "$(dirname  "${BASH_SOURCE[0]}")/../.." >/dev/null 2>&1 && pwd )"
```
```suggestion
HOMEgfs="$(cd "$(dirname  "${BASH_SOURCE[0]}")/../.." >/dev/null 2>&1 && pwd )"
```

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -0,0 +1,70 @@
+#  Set up runtime environment varibles for accounts on supproted machines
+#########################################################################
+
+source "${HOMEGFS_DIR}/ush/detect_machine.sh"
```
```suggestion
source "${HOMEgfs}/ush/detect_machine.sh"
```

Might as well also call `module-setup.sh` here.

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -0,0 +1,70 @@
+case ${MACHINE_ID} in
+  hera | orion)
+   echo "Running Automated Testing on ${MACHINE_ID}"
+   source "${HOMEGFS_DIR}/ci/environments/${MACHINE_ID}.sh"
```
```suggestion
   source "${HOMEgfs}/ci/environments/${MACHINE_ID}.sh"
```

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -0,0 +1,70 @@
+   ;;
+esac
+
+module use "${HOMEGFS_DIR}/modulefiles"
```
```suggestion
module use "${HOMEgfs}/modulefiles"
```

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -0,0 +1,70 @@
+#!/bin/bash
+#set -eux
```
```suggestion
set -eux
```

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -0,0 +1,70 @@
+   exit 1
+   ;;
+esac
+
```
disable garbage from `module` commands.
```suggestion
set +x
```

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -0,0 +1,70 @@
+module use "${HOMEGFS_DIR}/modulefiles"
+module load "module_gwsetup.${MACHINE_ID}"
+module list
+
```
re-enable trace
```suggestion
set -x
```

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -0,0 +1,70 @@
+  pr_dir="${GFS_CI_ROOT}/PR/${pr}"
+  for experiments in "${pr_dir}/RUNTESTS/"*; do
+    pslot=$(basename "${experiments}")
+    xml="${pr_dir}/RUNTESTS/${pslot}/EXPERMENT/${pslot}/${pslot}.xml"
```
```suggestion
    xml="${pr_dir}/RUNTESTS/${pslot}/EXPERIMENT/${pslot}/${pslot}.xml"
```

TerrenceMcGuinness-NOAA replied:
Ugh.

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -0,0 +1,70 @@
+  for experiments in "${pr_dir}/RUNTESTS/"*; do
+    pslot=$(basename "${experiments}")
+    xml="${pr_dir}/RUNTESTS/${pslot}/EXPERMENT/${pslot}/${pslot}.xml"
+    db="${pr_dir}/RUNTESTS/${pslot}/EXPERMENT/${pslot}/${pslot}.db"
```
```suggestion
    db="${pr_dir}/RUNTESTS/${pslot}/EXPERIMENT/${pslot}/${pslot}.db"
```

TerrenceMcGuinness-NOAA replied:
yep, same as other.

aerorahul commented on ci/scripts/run_ci.sh:
```diff
@@ -0,0 +1,70 @@
+    xml="${pr_dir}/RUNTESTS/${pslot}/EXPERMENT/${pslot}/${pslot}.xml"
+    db="${pr_dir}/RUNTESTS/${pslot}/EXPERMENT/${pslot}/${pslot}.db"
+    echo "Running: ${rocotorun} -v 6 -w ${xml} -d ${db}"
+    "${rocotorun}" -v 6 -w "${xml}" -d "${db}"
```
Why `-v 6`?  Why not 0 or 10?

TerrenceMcGuinness-NOAA replied:
Does it go up that high? Ok.

aerorahul replied:
It has a log base 10 level scaling.
0 - no verbosity
10 - some
100 - some more
1000 - a lot more.

aerorahul commented on modulefiles/module_gwci.hera.lua:
```diff
@@ -0,0 +1,15 @@
+help([[
```
I will ignore these `module_gwci.MACHINE.lua` files since we are not ready to do comparisons yet.

TerrenceMcGuinness-NOAA replied:
Ok, yes, agreed. Not really this PR.

aerorahul commented on test/diff_grib_files.py:
```diff
@@ -66,10 +66,17 @@ def count_nonid_corr(test_string: str, quiet=False):
 if __name__ == '__main__':
-    fileA = sys.argv[0]
-    fileB = sys.argv[1]
+    fileA = sys.argv[1]
```
When/where did this come from?

TerrenceMcGuinness-NOAA replied:
It's bug in Walter's code. **argv[0]** is the script base name and he intended it to pick up the first argument.

aerorahul reviewed (changes requested):
Mostly looks good.
Some minor comments and some exclusions to hold for subsequent work.

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/clone-build_ci.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/run_ci.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/experiments/C96_atm3DVar.yaml

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/clone-build_ci.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/run_ci.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Apply suggestions from code review

Used traditional var name for HOMEgfs (that I changed that)

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/clone-build_ci.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

aerorahul commented on ci/scripts/clone-build_ci.sh:
```diff
@@ -73,37 +68,53 @@ if [[ "${pr_state}" != "OPEN" ]]; then
 # run build script
-cd sorc
+cd sorc || exit 1
 export BUILD_JOBS=8
 rm -rf log.build
```
Can you please move these 2 lines right before the call to `./build_all.sh`?  It makes sense there.

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver.sh

removed export on MACHINE_ID as it is not needed

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/run_ci.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/run_ci.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/run_ci.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
added module-setup.sh in place of module purge

TerrenceMcGuinness-NOAA committed:
couple more suggestions, fix build_status off of link and more build settings next to the build script

TerrenceMcGuinness-NOAA committed:
just moved an export to where it is more approprate

TerrenceMcGuinness-NOAA committed:
improved symantics by replacing environments with platform and experments with cases

TerrenceMcGuinness-NOAA committed:
Update diff_grib_files.py

aerorahul commented on ci/cases/C96_atm3DVar_also.yaml:
```diff
@@ -0,0 +1,14 @@
+experiment:
```
Instead of this, you could retain the one that is deleted.
`C96C48_hybatmDA.yaml`

TerrenceMcGuinness-NOAA replied:
ok

TerrenceMcGuinness-NOAA committed:
Delete C96_atm3DVar_also.yaml

TerrenceMcGuinness-NOAA committed:
added second case C96C48-hyb-ATM-DA

aerorahul reviewed (approved):
looks good.

TerrenceMcGuinness-NOAA committed:
Update run_ci.sh

rocotorun verbose settings are 1,10,100 ect ..

WalterKolczynski-NOAA commented on ci/scripts/clone-build_ci.sh:
```diff
@@ -45,62 +40,79 @@ while getopts "p:d:o:h" opt; do
+  exit 1
+fi  
+
+cd "${repodir}" || exit
```
This should be a non-zero exit

WalterKolczynski-NOAA replied:
This was marked as resolved but not actually resolved.

TerrenceMcGuinness-NOAA replied:
I double checked in my branch at it was but not here.  Let me do that again and see why.

TerrenceMcGuinness-NOAA replied:
Yeah it has been updated.  It is just not being reflected here in the chat stream.  Follow this [link](https://github.com/NOAA-EMC/global-workflow/blob/2d365727bba5a90ec81372bebd43b14be3f1262f/ci/scripts/clone-build_ci.sh#L50).

TerrenceMcGuinness-NOAA replied:
Ok, there it is.  I think going back to my repo and syncing it to the main repo updated these changes upstream.

WalterKolczynski-NOAA replied:
Still no exit code.

TerrenceMcGuinness-NOAA replied:
![Screenshot 2023-04-21 at 2 09 47 PM](https://user-images.githubusercontent.com/66784684/233705801-5ef71e20-b412-40b0-8f09-d13119f46deb.png)

TerrenceMcGuinness-NOAA replied:
Line 50 has had that `exit 1` every time I check.

WalterKolczynski-NOAA commented on modulefiles/module_gwci.orion.lua:
```diff
@@ -0,0 +1,20 @@
+Load environment to run GFS workflow ci scripts on Orion
+]])
+
+-- Temporary until official hpc-stack is updated
```
What's wrong with hpc-stack on Orion? It has all these.

WalterKolczynski-NOAA replied:
I think this comment got separated from where it belongs before loading my stack to get the pyenv.

WalterKolczynski-NOAA commented on modulefiles/module_gwci.hera.lua:
```diff
@@ -0,0 +1,15 @@
+load(pathJoin("hpc-intel", "18.0.5.274"))
+load(pathJoin("hpc-impi", "2018.0.4"))
+
+load(pathJoin("netcdf","4.7.0"))
```
```suggestion
load(pathJoin("netcdf","4.7.4"))
```

This is the version used by workflow.

WalterKolczynski-NOAA replied:
Pretty sure you had this fixed at one point and now it has reverted back.

WalterKolczynski-NOAA commented on modulefiles/module_gwci.orion.lua:
```diff
@@ -0,0 +1,20 @@
+
+prepend_path("MODULEPATH", "/apps/modulefiles/core")
+load(pathJoin("netcdf","/4.7.2"))
+load(pathJoin("netcdf","4.7.2-parallel"))
```
```suggestion
load(pathJoin("netcdf","4.7.4"))
```

WalterKolczynski-NOAA commented on modulefiles/module_gwci.orion.lua:
```diff
@@ -0,0 +1,20 @@
+
+-- Temporary until official hpc-stack is updated
+
+prepend_path("MODULEPATH", "/apps/modulefiles/core")
```
Should be using the Orion hpc-stack just like Hera:
```suggestion
prepend_path("MODULEPATH", "/apps/contrib/NCEP/hpc-stack/libs/hpc-stack/modulefiles/stack")

load(pathJoin("hpc", "1.1.0"))
load(pathJoin("hpc-intel", "2018.4"))
load(pathJoin("hpc-impi", "2018.4"))
```

TerrenceMcGuinness-NOAA committed:
fixed fail logic in driver for setupexpt and bug with path in yaml for second case

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature_1428_CI-cronjobs' of github.com:TerrenceMcGuinness-NOAA/global-workflow into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
fixed updated path to EXPDIR for staged test cases in run_ci script

TerrenceMcGuinness-NOAA committed:
Added additional script to check to see if experments have completed succesfully

TerrenceMcGuinness-NOAA committed:
added new label stage to include CI--Building before Running

TerrenceMcGuinness-NOAA committed:
moved check for open PR after it is cloned

TerrenceMcGuinness-NOAA committed:
updated old repo_url varible name in check script

TerrenceMcGuinness-NOAA committed:
added comment to GitHub when case completes

TerrenceMcGuinness-NOAA committed:
added cool SHELLCHECK suggestiongs (amazing tool)

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature_1428_CI-cronjobs' of github.com:TerrenceMcGuinness-NOAA/global-workflow into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
Update check_ci.sh

TerrenceMcGuinness-NOAA committed:
Update check_ci.sh

TerrenceMcGuinness-NOAA committed:
added a could of shellcheck suggestions for || true

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature_1428_CI-cronjobs' of github.com:TerrenceMcGuinness-NOAA/global-workflow into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
Update module_gwci.orion.lua

TerrenceMcGuinness-NOAA committed:
Update modulefiles/module_gwci.orion.lua

better set of netcdf loads with parallel build

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update modulefiles/module_gwci.orion.lua

Orion hpc-stack just like Hera

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update modulefiles/module_gwci.hera.lua

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
added set -e at top of BASH scripts and then only disabled it when we do not want to break out of checks

TerrenceMcGuinness-NOAA committed:
fixed login of label update in check_ci script and added parallel version of NetCDF for threaded nccmp

TerrenceMcGuinness-NOAA committed:
improved login for success of PR with check_ci script

TerrenceMcGuinness-NOAA committed:
forgot to add removing of PR from list on successfully completing all cases

TerrenceMcGuinness-NOAA committed:
forgot to add removing of PR from list on successfully completing all cases

TerrenceMcGuinness-NOAA committed:
fixed bug with counting remaining cases

TerrenceMcGuinness-NOAA committed:
added || true to BASH command for shellnorms in check ci script

TerrenceMcGuinness-NOAA committed:
fixed race condition for check_ci script thinking success before experments are created (checks for RUNTESTS dir)

WalterKolczynski-NOAA commented on ci/scripts/check_ci.sh:
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-
+set -aux
```
Was this supposed to be -eux?

TerrenceMcGuinness-NOAA replied:
oops .. yes

TerrenceMcGuinness-NOAA replied:
I don't know why in this window it is not updated, but if you press edit you will see it is now `set -eux`

WalterKolczynski-NOAA commented on ci/scripts/driver.sh:
```diff
@@ -48,17 +48,19 @@ esac
 ############################################################
 pr_list_file="open_pr_list"
-rm -f "${pr_list_file}"
+touch "${GFS_CI_ROOT}/${pr_list_file}"
```
If this isn't being removed and two lines down you are just appending, how will old PRs ever be removed?

TerrenceMcGuinness-NOAA replied:
Good one!  That should have been done in `check_ci.sh`, but you are right, currently it only removes the cases and then flags success with all cases are gone.  It then "removes" the PR from the list, but all the dirs are left, which they should not on success.  I'll go back and update `check_ci.sh` to then also remove rest of the PR if all cases pass.

TerrenceMcGuinness-NOAA replied:
Done

TerrenceMcGuinness-NOAA replied:
OK, now the check script will remove all cases that succeed.  When the cron check script has zero cases left under `../${PR}/RUNTESTS` (when the **RUNTESTS** dir is present)  the script will completely delete the PR and remove it from the list.  This works because the **RUNTESTS** dir is not created until the cases are setup.

WalterKolczynski-NOAA commented on modulefiles/module_gwci.orion.lua:
```diff
@@ -7,7 +7,7 @@ prepend_path("MODULEPATH", "/apps/contrib/NCEP/hpc-stack/libs/hpc-stack/modulefi
 load(pathJoin("hpc-intel", "2018.4"))
 load(pathJoin("hpc-impi", "2018.4"))
-load(pathJoin("netcdf","4.7.4"))
+load(pathJoin("netcdf","4.7.2-parallel"))
```
Why not 4.7.4 parallel?

Also, check with NCEPLIBS, but I thought when impi is loaded, the netcdf version already supports parallel.

TerrenceMcGuinness-NOAA replied:
module load complained that **nccmp** needed Parallel NetCDF 4.7.2 and would not load it, not sure about 4.7.4, I will check this again tomorrow with you suggestion and see about the versions.

TerrenceMcGuinness-NOAA replied:
I recall `ldd` showed that the runtime libs where not in place for **nccmp** when loading iMPI and NetCDF during my choice to go with Parallel 4.7.2, I'll check again and see, maybe I had the wrong Intel libs.

TerrenceMcGuinness-NOAA replied:
```
(gfs_workflow) mterry (Orion-login-1) modulefiles (feature_1428_CI-cronjobs) $ ldd /apps/nccmp-1.8.5/bin/nccmp
	linux-vdso.so.1 =>  (0x00007ffdeb1fa000)
	libnetcdf.so.15 => /apps/intel-2018/netcdf-4.7.2-parallel/lib/libnetcdf.so.15 (0x00007f16df786000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f16df484000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f16df268000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f16dee9b000)
	libpnetcdf.so.4 => /apps/intel-2018/pnetcdf-1.12.0/lib/libpnetcdf.so.4 (0x00007f16de33f000)
	libifport.so.5 => /apps/intel-2018/intel-2018.u4/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libifport.so.5 (0x00007f16de113000)
	libifcoremt.so.5 => /apps/intel-2018/intel-2018.u4/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libifcoremt.so.5 (0x00007f16ddd7f000)
	libirc.so => /apps/intel-2018/intel-2018.u4/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libirc.so (0x00007f16ddb11000)
	libhdf5_hl.so.100 => /apps/intel-2018/hdf5-1.10.5-parallel/lib/libhdf5_hl.so.100 (0x00007f16dd8e8000)
	libhdf5.so.103 => /apps/intel-2018/hdf5-1.10.5-parallel/lib/libhdf5.so.103 (0x00007f16dd212000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f16dd00e000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f16dcdf8000)
	libcurl.so.4 => /lib64/libcurl.so.4 (0x00007f16dcb8f000)
	libmpifort.so.12 => /apps/intel-2018/intel-2018.u4/compilers_and_libraries_2018.5.274/linux/mpi/intel64/lib/libmpifort.so.12 (0x00007f16dc7e6000)
	libmpi.so.12 => /apps/intel-2018/intel-2018.u4/compilers_and_libraries_2018.5.274/linux/mpi/intel64/lib/release/libmpi.so.12 (0x00007f16dbb8a000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f16db982000)
	libimf.so => /apps/intel-2018/intel-2018.u4/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libimf.so (0x00007f16db3ef000)
	libsvml.so => /apps/intel-2018/intel-2018.u4/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libsvml.so (0x00007f16d9ae1000)
	libirng.so => /apps/intel-2018/intel-2018.u4/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libirng.so (0x00007f16d976d000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f16d9557000)
	libintlc.so.5 => /apps/intel-2018/intel-2018.u4/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libintlc.so.5 (0x00007f16d92e9000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f16dfc9c000)
	libmpicxx.so.12 => /apps/intel-2018/intel-2018.u4/compilers_and_libraries_2018.5.274/linux/mpi/intel64/lib/libmpicxx.so.12 (0x00007f16d90c9000)
	libcilkrts.so.5 => /apps/intel-2018/intel-2018.u4/compilers_and_libraries_2018.5.274/linux/compiler/lib/intel64/libcilkrts.so.5 (0x00007f16d8e8c000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f16d8b85000)
	libidn.so.11 => /lib64/libidn.so.11 (0x00007f16d8952000)
	libssh2.so.1 => /lib64/libssh2.so.1 (0x00007f16d8728000)
	libssl3.so => /lib64/libssl3.so (0x00007f16d84d6000)
	libsmime3.so => /lib64/libsmime3.so (0x00007f16d82af000)
	libnss3.so => /lib64/libnss3.so (0x00007f16d7f82000)
	libnssutil3.so => /lib64/libnssutil3.so (0x00007f16d7d52000)
	libplds4.so => /lib64/libplds4.so (0x00007f16d7b4e000)
	libplc4.so => /lib64/libplc4.so (0x00007f16d7949000)
	libnspr4.so => /lib64/libnspr4.so (0x00007f16d770b000)
	libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2 (0x00007f16d74be000)
	libkrb5.so.3 => /lib64/libkrb5.so.3 (0x00007f16d71d5000)
	libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x00007f16d6fa2000)
	libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007f16d6d9e000)
	liblber-2.4.so.2 => /lib64/liblber-2.4.so.2 (0x00007f16d6b8f000)
	libldap-2.4.so.2 => /lib64/libldap-2.4.so.2 (0x00007f16d693a000)
	libssl.so.10 => /lib64/libssl.so.10 (0x00007f16d66c8000)
	libcrypto.so.10 => /lib64/libcrypto.so.10 (0x00007f16d6266000)
	libkrb5support.so.0 => /lib64/libkrb5support.so.0 (0x00007f16d6056000)
	libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x00007f16d5e52000)
	libresolv.so.2 => /lib64/libresolv.so.2 (0x00007f16d5c39000)
	libsasl2.so.3 => /lib64/libsasl2.so.3 (0x00007f16d5a1c000)
	libselinux.so.1 => /lib64/libselinux.so.1 (0x00007f16d57f5000)
	libcrypt.so.1 => /lib64/libcrypt.so.1 (0x00007f16d55be000)
	libpcre.so.1 => /lib64/libpcre.so.1 (0x00007f16d535c000)
	libfreebl3.so => /lib64/libfreebl3.so (0x00007f16d5159000)
(gfs_workflow) mterry (Orion-login-1) modulefiles (feature_1428_CI-cronjobs) $ module list

Currently Loaded Modules:
  1) hdf5/1.10.5-parallel   3) netcdf/4.7.2-parallel   5) contrib/0.1    7) hpc/1.2.0      9) hpc-intel/2018.4   11) hpc-miniconda3/4.6.14  13) module_gwci.orion  15) hpc-impi/2018.4
  2) pnetcdf/1.12.0         4) nccmp/1.8.5             6) wgrib2/3.0.2   8) intel/2018.4  10) miniconda3/4.6.14  12) gfs_workflow/1.0.0     14) impi/2018.4

 

(gfs_workflow) mterry (Orion-login-1) modulefiles (feature_1428_CI-cronjobs) $ module unload netcdf/4.7.2-parallel
(gfs_workflow) mterry (Orion-login-1) modulefiles (feature_1428_CI-cronjobs) $ module load netcdf/4.7.4

The following have been reloaded with a version change:
  1) hdf5/1.10.5-parallel => hdf5/1.10.6

(gfs_workflow) mterry (Orion-login-1) modulefiles (feature_1428_CI-cronjobs) $ ldd /apps/nccmp-1.8.5/bin/nccmp
	linux-vdso.so.1 =>  (0x00007ffe86b9d000)
	libnetcdf.so.15 => not found
	libm.so.6 => /lib64/libm.so.6 (0x00007ff44ebf2000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007ff44e9d6000)
	libc.so.6 => /lib64/libc.so.6 (0x00007ff44e609000)
	/lib64/ld-linux-x86-64.so.2 (0x00007ff44eef4000)
(gfs_workflow) mterry (Orion-login-1) modulefiles (feature_1428_CI-cronjobs) $ 

```

WalterKolczynski-NOAA replied:
After talking with Terry, nccmp is the issue (it has incorrect prerequisites on Orion). Moving to `nccmp/1.8.7.0` should fix things, as that module has the correct prerequisites. Hera should also be updated for consistency.

TerrenceMcGuinness-NOAA replied:
Walter figured out that nccmp 1.8.7.0 would link ok with intel MPI on Orion

TerrenceMcGuinness-NOAA committed:
Upates to final success check with multiple cases in check_ci ie num_cases=0

TerrenceMcGuinness-NOAA committed:
Update check_ci.sh

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA committed:
Update clone-build_ci.sh

TerrenceMcGuinness-NOAA committed:
couple of bug updates for check script when deleting successfull PRs, fixed issue with nccmp linking on Orion

TerrenceMcGuinness-NOAA reviewed (commented):
Responded to all comments by the Code Manager

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature_1428_CI-cronjobs' of github.com:TerrenceMcGuinness-NOAA/global-workflow into feature_1428_CI-cronjobs

TerrenceMcGuinness-NOAA reviewed (commented):
had to do another push ..

TerrenceMcGuinness-NOAA committed:
Update check_ci.sh

Added "" for BASH pep and needed to send message back to GitHub before we delete the PR for good.

TerrenceMcGuinness-NOAA reviewed (commented):
Needed one last commit for BASH PEP and a missing message.

TerrenceMcGuinness-NOAA committed:
updated nccmp 1.8.7.0 and fixed a shellnorm typo to close quote

TerrenceMcGuinness-NOAA reviewed (commented):
last fix to nccmp on Hera and a close quote on a typo for shellnorm

TerrenceMcGuinness-NOAA reviewed (commented):
ready to merge
